docs/internal/generics.md

```diff
@@ -156,3 +156,110 @@ let z : U32 & imm = obj.convert(y); // ERROR: Not able to define what type y is
+
+The end result would be:
+```ts
+// Definition of monomorphic classes only
+class Generic_U32 { field : U32 & imm; ... };
+class Generic_MyClass { field : MyClass & imm; ... }
+
+// Direct use of the concrete classes
+let foo : Generic_U32 = ...;
+let bar : Generic_MyClass = ...;
+```
+
+Note that type unions (`A | B`) are still monomorphic types.
+To resolve which implementation works with each inner type (`A` or `B`) one must use `match` inside the method.
+
+Example:
+```ts
+class Generic[A] {
+  field : A & imm;
+  // Implement constructor with match on types
+  add(other : A & imm) {
+    match (A)
+    { Numeric => field + other }
+    { String => field.append(other) }
+    { _ => Throw[None] } // Or some better error handling
+  }
+}
+...
+// This creates a monomorphic union type
+let foo : Generic[U32 | String] = ...;
+
+// This creates a monomorphic single type
+let bar : Generic[U32] = ...;
+
+// This union of Numeric types isn't quite static
+let baz : Generic[U32 | F64] = ...;
+```
+
+On the first case, `foo` is a monomorphic union between `U32` and `String`, which means, at run-time, either can be passed to `add` and we can't decide at compile time.
+So the implementation retains the `match` and will check if the types are `Numeric` or `String`.
+Note that both `field` and `other` must have the same type (both are `A`).
+
+However, on the second case, `bar` is a compile-time single monomorphic type, which guarantees that it will always implement the `Numeric` interface and therefore there's only one `match` statement that fits.
+In this case, the compiler is free to elide the `match` altogether and emit only the case where the type is `Numeric`, ie. `field + other`.
+In this case, the type is guaranteed to be `U32` to the call to `+` is a static call to `U32::+(U32, U32)`.
+
+Finally, in the last case, `baz`'s types are both `Numeric`, but the `+` operator for each is different.
+So, while the compiler is free to elide all other `match` cases, the dispatch of the `+` method is dynamic, ie. `field.+(other)`.
+
+The code above results in the following compile-time concrete types:
+```ts
+// (U32 | String) need different initialisation and add match
+class Generic_U32_String {
+  field : (U32 | String) & imm;
+  // Implement constructor with match on types
+  add(other : (U32 | String) & imm) {
+    match (typeof(field))
+    { U32 => U32::+(field, other) }
+    { String => String::append(field, other) }
+    // No need for error handling 
+  }
+}
+
+// Just U32 elides everything
+class Generic_U32 {
+  field : U32 & imm;
+  // Implement constructor for U32 only
+  add(other : U32 & imm) {
+    U32::+(field, other)
+  }
+}
+
+// Both numeric, need dynamic dispatch
+class Generic_U32_F64 {
+  field : (U32 | F64) & imm;
+  // Implement constructor with match on types
+  add(other : (U32 | F64) & imm) {
+    field.+(other);
```

Comment:
Nope, this `+` doesn't exist in either of the concrete types.

Reply:
See my `Self` comment.

In cases like this, where both types are numeric and both have `+` implemented (on their own types), how can I make it so that I don't need to match against each? Or is that impossible?

Reply:
If they both implement a `+` whose arguments are compatible with all of the types of the parameters that you pass, you don't need a `match`.  If they both implement `+` with argument types that are compatible with the concrete types that you have but not with the complete interface type that you're defining, then you need a match.

The problem here is that the concrete types *don't* implement a method that's compatible with your concrete types.  You example allows for `U32+F64` and `F64+U32` as the concrete types and neither is a defined operation.

Reply:
Right, perhaps I got the wrong types to be presenting this concept.

I didn't want to create completely new types that weren't familiar to readers because that's hard to connect.

I'll see if I can conjure up something better.